torch/distributed/_state_dict_utils.py

```diff
@@ -50,60 +60,97 @@ def _all_gather_sharded_tensor(
     return tensor
 
 
+class CompanionMismatch(Exception):
```

Comment:
Any reason we can't use a value error here?

Reply:
We need a customized exception to distinguish mismatched error from the actual error. Since the mismatched error is not an actual error, we will silently ignore the error. So if we use ValueError, we may accidentally ignore a real error.